File: SimpleDocbookEditor/ui/MainWindow.py
# MainWindow.py
# -*- coding=utf-8 -*-
# -*- Mode: Python; indent-tabs-mode: nil; tab-width: 4; coding: utf-8 -*-
#
# Copyright © 2014 Gwendal Le Bihan
# 
# This file is part of Simple DocBook Editor.
# 
# Simple DocBook Editor is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# Simple DocBook Editor is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Simple DocBook Editor.  If not, see <http://www.gnu.org/licenses/>.

import gtk
import webkit
import os
import urlparse
import urllib
import json
import logging
from OpenBookDialog import OpenBookDialog
from ImageBrowserDialog import ImageBrowserDialog
from AboutDialog import AboutDialog
from SaveQuitDialog import SaveQuitDialog
from ..informations import *

class MainWindow(object):
    def __init__(self, application):
        self._application = application
        
        builder = gtk.Builder()
        builder.add_from_file(os.path.join(self._application.cli_options.share_dir, UNIX_APPNAME, "ui", "ui.glade"))
        self._window = builder.get_object("main_window")
        self._window.add_accel_group(builder.get_object("main_accelgroup"))
        
        self._window.set_title(_(APPNAME))
        
        self._window.set_size_request(self._application.config["UI/width"], self._application.config["UI/height"])
        self._maximized = self._application.config["UI/maximized"]
        if self._maximized:
            self._window.maximize()
        
        self._webview = webkit.WebView()
        builder.get_object("webview_container").add(self._webview)
        self._webview.get_settings().set_property('enable-file-access-from-file-uris', 1)
        self._webview.get_settings().set_property('enable-default-context-menu', 0)
        self._webview.connect("script-alert", self._on_webview_script_alert)
        self._webview_load_finished = False
        self._webview_pending_commands = []
        
        self._statusbar = builder.get_object("statusbar")
        
        self._load_main_window()
        
        self._window.connect("delete_event", self._on_window_delete_event)
        builder.get_object("quit_menu_item").connect("activate", self._on_quit_clicked)
        builder.get_object("open_book_menuitem").connect("activate", self._on_open_book_clicked)
        builder.get_object("save_book_menuitem").connect("activate", self._on_save_book_clicked)
        builder.get_object("close_book_menuitem").connect("activate", self._on_close_book_clicked)
        builder.get_object("about_menuitem").connect("activate", self._on_about_clicked)
        
        self._open_book_dialog = OpenBookDialog(self._application, self._window)
        self._about_dialog = AboutDialog(self._window)
        self._save_quit_dialog = SaveQuitDialog(self._window)
        self._image_browser_dialog = ImageBrowserDialog(self._application, self._window)
        
        self._mapped = False
        self._window.connect("size-allocate", self._on_size_allocate)
        self._window.connect("window-state-event", self._on_window_state_event)
        self._window.connect("map-event", self._on_map_event)
    
    def _on_map_event(self, window, event):
        self._mapped = True
        self._window.connect("size-allocate", self._on_size_allocate)
        self._window.connect("window-state-event", self._on_window_state_event)
    
    def _on_window_state_event(self, window, event):
        if self._mapped:
            self._maximized = (event.new_window_state & gtk.gdk.WINDOW_STATE_MAXIMIZED != 0)
    
    def _on_size_allocate(self, window, rectangle):
        if self._mapped:
            if self._maximized:
                self._application.config["UI/maximized"] = True
            else:
                self._application.config["UI/maximized"] = False
                width, height = self._window.get_size()
                self._application.config["UI/width"] = width
                self._application.config["UI/height"] = height
    
    def _check_quit(self):
        if self._application.book and self._application.book.changed:
            resp = self._save_quit_dialog.run()
            if resp == gtk.RESPONSE_YES:
                self._application.book.save()
            elif resp != gtk.RESPONSE_NO:
                return True
        gtk.main_quit()
        return False
    
    def _on_window_delete_event(self, window, event):
        return self._check_quit()
    
    def _on_close_book_clicked(self, menuitem):
        if self._application.book:
            if self._application.book.changed:
                resp = self._save_quit_dialog.run()
                if resp == gtk.RESPONSE_YES:
                    self._application.book.save()
                elif resp != gtk.RESPONSE_NO:
                    return
            self._application.unload_book()
            self.send_command("set_doc_structure(null)")
    
    def _on_about_clicked(self, menuitem):
        self._about_dialog.run()
        
    def _on_quit_clicked(self, menuitem):
        self._check_quit()
    
    def _on_open_book_clicked(self, menuitem):
        filename = self._open_book_dialog.run()
        if filename:
            if self._application.book:
                self._on_close_book_clicked(menuitem)
            if self._application.book:
                return
            self._application.load_book(filename)
    
    def _on_save_book_clicked(self, menuitem):
        self._application.book.save()
    
    def show_all(self):
        self._window.show_all()
    
    def _do_send_command(self, command):
        self._webview.execute_script(command)
    
    def send_command(self, command):
        if self._webview_load_finished:
            self._do_send_command(command)
        else:
            self._webview_pending_commands.append(command)
    
    def _on_webview_script_alert(self, editor, frame, message):
        logging.debug("_on_webview_script_alert:%s" % ":".join(message.split(":")[:2]))
        
        if ":" in message:
            i = message.index(":")
            command = message[:i]
            params = message[i+1:]
        else:
            command = message
            params = ""
        
        if command == "reload_doc_structure":
            if self._application.book:
                self.send_command("set_doc_structure(%s)" % json.dumps(self._application.book.get_structure_tree()))
        elif command == "editor_ready":
            while len(self._webview_pending_commands):
                command = self._webview_pending_commands[0]
                del self._webview_pending_commands[0]
                self._do_send_command(command)
            self._webview_load_finished = True
        elif command == "load_doc_section":
            section_id = int(params)
            section = self._application.book.find_section_by_id(section_id)
            xml = section.get_xml_text()
            if section.edit_mode == "html":
                self.send_command("set_edit_data(%s)" % json.dumps({"section_id": section_id, "edit_mode": section.edit_mode, "html": str(section.get_html()), "xml": str(xml)}))
            else:
                self.send_command("set_edit_data(%s)" % json.dumps({"section_id": section_id, "edit_mode": section.edit_mode, "xml": str(xml)}))
        elif command == "set_section_contents":
            i = params.index(":")
            section_id = int(params[:i])
            contents = params[i+1:]
            section = self._application.book.find_section_by_id(section_id)
            parsing_error = section.update_from_html(contents)
            if parsing_error:
                parsing_error_json = json.dumps(_("You have an error in your document. Please fix it before editing another section."))
            else:
                parsing_error_json = json.dumps(None)
            self.send_command("set_parsing_error(%s)" % parsing_error_json)
            self.send_command("set_doc_structure(%s)" % json.dumps(self._application.book.get_structure_tree()))
        elif command == "set_section_source_contents":
            i = params.index(":")
            section_id = int(params[:i])
            contents = params[i+1:]
            section = self._application.book.find_section_by_id(section_id)
            parsing_error = section.update_from_xml(contents)
            if parsing_error:
                parsing_error_json = json.dumps(_("You have an error in your document. Please fix it before editing another section."))
            else:
                parsing_error_json = json.dumps(None)
            self.send_command("set_parsing_error(%s)" % parsing_error_json)
            self.send_command("set_doc_structure(%s)" % json.dumps(self._application.book.get_structure_tree()))
        elif command == "browse_image":
            i = params.index(":")
            field_name = params[:i] 
            url = params[i+1:]
            if url:
                filename = urllib.url2pathname(url)[7:]
            else:
                filename = None
            new_filename = self._image_browser_dialog.run(filename)
            if new_filename:
                self.send_command("set_file_browser_filename(%s)" % json.dumps({"field_name": field_name, "url": urlparse.urljoin('file:', urllib.pathname2url(new_filename))}))
        elif command == "load_config":
            self._do_send_command("set_config(%s)" % json.dumps(self._application.config.serialize()))
        elif command == "translate":
            self.send_command("set_translation(%s)" % json.dumps({"mesg": params, "translation": _(params)}))
        else:
            return False
            
        return True
    
    def _load_main_window(self):
        filename = os.path.join(self._application.cli_options.share_dir, UNIX_APPNAME, "ui", "main_window.html")
        f = open(filename)
        data = f.read()
        f.close()
        
        while "_(" in data:
            i = data.index("_(")
            pcount = 1
            j = i + 2
            while pcount != 0:
                if data[j] == "(":
                    pcount += 1
                if data[j] == ")":
                    pcount -= 1
                j += 1
            data = data[:i] + _(data[i+2:j-1]) + data[j:]
        
        self._webview.load_html_string(data, urlparse.urljoin('file:', urllib.pathname2url(filename)))
    
    def refresh_view_for_new_book(self, section_id = 0):
        self.send_command("refresh_view_for_new_book(%d)" % section_id)
        if self._application.book:
            self._window.set_title("%s - %s" % (_(APPNAME), self._application.book.filename))
        else:
            self._window.set_title("%s" % _(APPNAME))

File: share/simple-docbook-editor/ui/main_window.js
/**
 * main_window.js
 *
 * Copyright © 2014 Gwendal Le Bihan
 * 
 * This file is part of simple-docbook-editor.
 * 
 * Simple DocBook Editor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Simple DocBook Editor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Simple DocBook Editor.  If not, see <http://www.gnu.org/licenses/>.
 */

var doc_structure_closed_nodes = new Array();
var selected_doc_section = null;
var edited_section_id = 0;
var edited_section_node = null;
var saveTimeout = null;
var sourceSaveTimeout = null;
var current_file_browser_window = null;
var source_editor = null;
var parsing_error = null;
var app_config = {};
var first_config_load = true;
var translations = {};

function _(mesg)
{
    if (!translations[mesg])
    {
        alert("translate:" + mesg);
    }
    console.log(mesg + ":" + translations[mesg]);
    return (translations[mesg] ? translations[mesg] : mesg);
}

function set_translation(data)
{
    translations[data.mesg] = data.translation;
}

function reload_doc_structure()
{
    alert("reload_doc_structure");
}

function set_doc_structure(doc_structure)
{
    if (doc_structure === null)
    {
        jQuery("#doc_structure").tree("loadData", []);
    }
    else
    {
        jQuery("#doc_structure").tree("loadData", [doc_structure]);
        if (selected_doc_section)
        {
            node = jQuery("#doc_structure").tree("getNodeById", selected_doc_section);
            jQuery("#doc_structure").tree('selectNode', node);
        }
        for (var i in doc_structure_closed_nodes)
        {
            node = jQuery("#doc_structure").tree("getNodeById", doc_structure_closed_nodes[i]);
            if (node)
            {
                jQuery("#doc_structure").tree('closeNode', node, false);
            }
        }
    }
}

function load_doc_section(section_id)
{
    if (saveTimeout)
    {
        clearTimeout(saveTimeout);
        saveTimeout = null;
    }
    alert("load_doc_section:" + section_id);
}

function set_edit_data(edit_data)
{
    if (edit_data.edit_mode == "html")
    {
        jQuery("#maintabs").tabs("enable", 0);
        edited_section_node = jQuery(edit_data.html);
        tinymce.get("tinymcecontainer").setContent(edited_section_node.html());
    }
    else
    {
        if (jQuery("#maintabs").tabs("option", "active") != 1)
        {
            jQuery("#maintabs").tabs("option", "active", 1);
        }
        jQuery("#maintabs").tabs("disable", 0);
    }
    if (edited_section_id != edit_data.section_id && edit_data.edit_mode == "html")
    {
        tinymce.get("tinymcecontainer").undoManager.clear();
    }
    edited_section_id = edit_data.section_id;
    selected_doc_section = edit_data.section_id;
    update_editor_height();
    source_editor.setValue(edit_data.xml);
}

function update_editor_height(){
    jQuery("#tinymcecontainer_ifr").css("height", (jQuery("#editor_inside_wrapper").height() - jQuery("#tinymcecontainer_ifr").offset().top) + "px");
    source_editor.setSize(null, jQuery("#editor_inside_wrapper").height() - jQuery("div.CodeMirror").offset().top);
}

function refresh_view_for_new_book(section_id)
{
    tinymce.get("tinymcecontainer").setContent("");
    doc_structure_closed_nodes = new Array();
    selected_doc_section = section_id;
    edited_section_id = 0;
    reload_doc_structure();
}

function do_save_editor_contents()
{
    edited_section_node.html(tinymce.get("tinymcecontainer").getContent());
    if (edited_section_id)
    {
        alert('set_section_contents:' + edited_section_id + ':' + edited_section_node[0].outerHTML);
    }
}

function save_editor_contents()
{
    if (saveTimeout)
    {
        clearTimeout(saveTimeout);
    }
    saveTimeout = setTimeout(do_save_editor_contents, 300);
}

function do_save_source_editor_contents()
{
    alert("set_section_source_contents:" + edited_section_id + ":" + source_editor.getValue());
}

function save_source_editor_contents()
{
    if (sourceSaveTimeout)
    {
        clearTimeout(sourceSaveTimeout);
    }
    sourceSaveTimeout = setTimeout(do_save_source_editor_contents, 300);
}

function set_file_browser_filename(data)
{
    current_file_browser_window.document.getElementById(data.field_name).value = data.url;
}

function set_parsing_error(error)
{
    parsing_error = error;
}

function show_alert(mesg)
{
    var dialog = jQuery("<div/>");
    dialog.html(mesg);
    var buttons = {};
    buttons[_("OK")] = function()
    {
        jQuery(this).dialog("close");
    };
    jQuery(dialog).dialog({
        title: _("Error"),
        buttons: buttons,
        modal: true
    });
}

function set_config(config)
{
    app_config = config;
    if (first_config_load)
    {
        first_config_load = false;
        
        jQuery('#doc_structure').tree(
        {
            data: [],
            autoOpen: true,
            dragAndDrop: false,
            onCanSelectNode: function(node)
            {
                if (parsing_error && node.id != edited_section_id)
                {
                    show_alert(parsing_error);
                    return false;
                }
                return (node.edit_mode != null);
            }
        });
        jQuery("#doc_structure").bind('tree.select', function(event)
        {
            if (event.node.id != edited_section_id)
            {
                selected_doc_section = event.node.id;
                load_doc_section(event.node.id);
            }
        });
        jQuery("#doc_structure").bind('tree.open', function(event)
        {
            var i = doc_structure_closed_nodes.indexOf(event.node.id);
            while (i != -1){
                doc_structure_closed_nodes.splice(i, 1);
                i = doc_structure_closed_nodes.indexOf(event.node.id);
            }
        });
        jQuery("#doc_structure").bind('tree.close', function(event)
        {
            doc_structure_closed_nodes.push(event.node.id);
        });
        
        jQuery("#leftbar").resizable(
        {
            handles: "e",
            resize: function(event, ui)
            {
                jQuery("#editor_wrapper").css("left", jQuery("#leftbar").outerWidth());
            }
        });
        
        tinymce.init(
        {
            selector: "#tinymcecontainer",
            setup: function(editor)
            {
                editor.on('click', function(e)
                {
                    if (jQuery(e.target).hasClass("mceNonEditable") && jQuery(e.target).hasClass("subsection") && parseInt(jQuery(e.target).attr("data-section-id")) > 0)
                    {
                        if (parsing_error)
                        {
                            show_alert(parsing_error);
                        }
                        else
                        {
                            selected_doc_section = parseInt(jQuery(e.target).attr("data-section-id"));
                            load_doc_section(parseInt(jQuery(e.target).attr("data-section-id")));
                            reload_doc_structure();
                        }
                    }
                });
                editor.on('change', function(e)
                {
                    if (jQuery("#maintabs").tabs("option", "active") == 0)
                    {
                        save_editor_contents();
                    }
                });
                editor.on('init', function(e)
                {
                    tinymce.get("tinymcecontainer").getBody().onkeyup = function(event)
                    {
                        save_editor_contents();
                    }
                    setTimeout(function()
                    {
                        update_editor_height();
                        alert("editor_ready");
                    }, 100);
                });
            },
            menubar: false,
            statusbar: false,
            plugins: ["code image noneditable docbook_subsections"],
            toolbar: "bold italic underline strikethrough | cut copy paste | bullist numlist | undo redo | removeformat subscript superscript | image | add_docbook_subsection | code",
            relative_urls: false,
            content_style: (
                'div.subsection {background-color: #B3B3B3; cursor: pointer; padding: 5px; margin-bottom: 10px;}'
            ),
            object_resizing: "img,table,sup.footnote",
            file_browser_callback: function(field_name, url, type, win)
            {
                current_file_browser_window = win;
                if (type == "image")
                {
                    alert("browse_image:" + field_name + ":" + url);
                }
            }
        });
        
        source_editor = CodeMirror.fromTextArea(document.getElementById("source_editor"),
        {
            mode: 'text/html',
            autoCloseTags: true,
            lineNumbers: true,
            indentUnit: 4,
            lineWrapping: app_config["Source Editor"].linewrapping
        });
        source_editor.on("change", function(editor, event)
        {
            if (jQuery("#maintabs").tabs("option", "active") == 1)
            {
                save_source_editor_contents();
            }
        });
        
        jQuery(window).resize(function(event)
        {
            update_editor_height();
        });
        
        jQuery("#maintabs").tabs(
        {
            activate: function(event)
            {
                update_editor_height();
                load_doc_section(edited_section_id);
            }
        });
    }
}

jQuery(document).ready(function()
{
    alert("load_config");
});
